wire-runtime/src/jvmMain/kotlin/com/squareup/wire/internal/reflection.kt

```diff
@@ -30,6 +30,7 @@ fun <M : Message<M, B>, B : Message.Builder<M, B>> createRuntimeMessageAdapter(
   messageType: Class<M>,
   typeUrl: String?,
   syntax: Syntax,
+  loader: ClassLoader,
```

Comment:
can we put a default param here? I'm not 100% we don't need one (even though it's within internal)

I think we need one

Reply:
@autonomousapps is it important to be able to pass a classloader that's different from `messageType`'s classloader? You do have some examples in the test changes where you pass the unit test class's classloader, rather than the message class's classloader. In those cases they are the same, but generally it seems more correct to pass the message class's classloader?

And if `loader` is always going to be the message class's classloader, maybe we can avoid adding the `ClassLoader` argument here and in a few other places? Or is the problem that, when a class is cast into `Class<Nothing>` (as done in wire-gson-support) we lose the ability to grab its classloader?

Reply:
Great questions @bendotli . The truth is that I haven't quite settled on the best approach yet, which is why I left this PR in draft. I first wanted to ensure that wire's maintainers were ok with the general idea before going deeper.

I think having a fully injectable classloader would be useful, _however_ it is complicated by a couple of factors:
1. Our path to `Class.forName(String)` actually goes through Retrofit and a gson adapter, so we'd have to modify at least two other libraries for that level of control (I think).
2. If we provided an overloaded constructor in a few places, the primary constructor would need to handle the current case (no classloader). We can't let `null` be the default value, since that actually has special meaning on the JVM: it means "use the boot classloader." So we'd need some sentinel value, or maybe `Optional<ClassLoader>` (which I think is still a nonstarter in Android), or grab the classloader from another of the values passed into that constructor. The latter is what I've gone with in this draft.